### PR TITLE
Switch internals to use `&[u8]` and API to take `AsRef<[u8]>`.

### LINKED ISSUE #33: Add `from_utf8` and handle `AsRef<[u8]>` in all API calls
We currently ask for `&str` to be passed to all `LanguageIdentifier` methods, but we also perform a full parsing of the input when creating `TinyStr`.

This means that we could allow for `AsRef<[u8]>` instead, once https://github.com/zbraniecki/tinystr/pull/11 lands.

### PR DESCRIPTION
**Pull request by zbraniecki**
Fixes #33.

### CODE REVIEW

File: unic-langid-impl/src/subtags.rs [hunk 1]
```diff
@@ -18,56 +18,57 @@ pub fn parse_language_subtag(subtag: &str) -> Result<Option<TinyStr8>, ParserErr
...
             Ok(s.to_ascii_uppercase())
         }
         3 => {
-            if subtag.contains(|c: char| !c.is_ascii_digit()) {
+            let s = TinyStr4::from_utf8(subtag.as_ref()).map_err(|_| ParserError::InvalidSubtag)?;
+            if !s.is_ascii_numeric() {
                 return Err(ParserError::InvalidSubtag);
             }
-            Ok(subtag.parse().unwrap())
+            Ok(s)
```
Comment (zbraniecki) [inline unic-langid-impl/src/subtags.rs]: whopsie! That was an error of course, and I fixed it here.

**Comment by zbraniecki** [comment]
@emilio - similar request here. You can skip the first patch as it just separates out the `clear_*` methods.

The second aims to make our API handle `AsRef<[u8]>` in all public methods and converge on `&[u8]` internally in preparation for the integration into Gecko.

**Comment by zbraniecki** [comment]
Some nice perf wins:

```
language_identifier_construct/from_str
                        time:   [430.08 ns 432.05 ns 434.14 ns]
                        change: [-10.615% -7.3840% -3.7992%] (p = 0.00 < 0.05)
                        Performance has improved.

language_identifier_parser
                        time:   [362.10 ns 363.81 ns 365.52 ns]
                        change: [-11.820% -8.0448% -4.5286%] (p = 0.00 < 0.05)
                        Performance has improved.

language_identifier_parser_casing
                        time:   [350.00 ns 351.99 ns 353.91 ns]
                        change: [-14.341% -10.828% -7.3207%] (p = 0.00 < 0.05)
                        Performance has improved.

locale_canonicalize     time:   [7.7733 us 7.8349 us 7.9000 us]
                        change: [-20.504% -14.918% -9.2806%] (p = 0.00 < 0.05)
                        Performance has improved.

locale_parser           time:   [3.2635 us 3.3231 us 3.3889 us]
                        change: [-32.104% -28.488% -24.761%] (p = 0.00 < 0.05)
                        Performance has improved.
```

File: unic-langid-impl/src/lib.rs [hunk 2]
```diff
@@ -94,7 +111,7 @@ impl LanguageIdentifier {
     ///
     /// assert_eq!(li.to_string(), "fr-CA");
     /// ```
-    pub fn from_parts<S: AsRef<str>>(
+    pub fn from_parts<S: AsRef<[u8]>>(
```
Comment (emilio) [inline unic-langid-impl/src/lib.rs]: Hmm, this means that using `from_parts(None, None, None, &[])` would not compile and you'd need to specify the type in at least one, but I guess as long as there's a `::default()` that does it you're good.

File: unic-langid-impl/tests/language_identifier_test.rs [hunk 3]
```diff
@@ -104,35 +104,27 @@ fn test_set_fields() {
...
-    langid
-        .set_language(Some("pl"))
-        .expect("Setting language failed");
+    langid.set_language("pl").expect("Setting language failed");
     assert_eq!(&langid.to_string(), "pl");
 
-    langid
-        .set_language(Some("de"))
-        .expect("Setting language failed");
+    langid.set_language("de").expect("Setting language failed");
```
Comment (emilio) [inline unic-langid-impl/tests/language_identifier_test.rs]: This looks actually neater than the `Option`-based API IMO.

File: unic-locale-impl/src/extensions/mod.rs [hunk 4]
```diff
@@ -58,24 +57,30 @@ pub struct ExtensionsMap {
...
         while let Some(subtag) = st {
-            let subtag = subtag.to_ascii_lowercase();
+            if subtag.is_empty() {
+                break;
+            }
 
-            match subtag.as_str() {
-                "" => break,
-                "u" => {
+            match ExtensionType::from_byte(subtag[0]) {
```
Comment (emilio) [inline unic-locale-impl/src/extensions/mod.rs]: What guarantees that `subtag.len()` is one?

File: unic-locale-impl/src/extensions/transform.rs [hunk 5]
```diff
@@ -59,11 +56,15 @@ impl TransformExtensionList {
         Ok(())
     }
 
-    pub fn set_tfield(&mut self, tkey: &str, tvalue: Vec<&str>) -> Result<(), LocaleError> {
-        let tkey = parse_tkey(tkey)?;
+    pub fn set_tfield<S: AsRef<[u8]>>(
+        &mut self,
+        tkey: S,
+        tvalue: Vec<S>,
```
Comment (emilio) [inline unic-locale-impl/src/extensions/transform.rs]: Do you really need a vec? Would this suffice with a `&[S]`? seems like you throw the vec away anyways.

File: unic-locale-impl/src/extensions/unicode.rs [hunk 6]
```diff
@@ -43,36 +40,40 @@ fn parse_type(t: &str) -> Result<Option<TinyStr8>, ParserError> {
...
     pub fn is_empty(&self) -> bool {
         self.keywords.is_empty() && self.attributes.is_empty()
     }
 
-    pub fn set_keyword(&mut self, key: &str, value: Vec<&str>) -> Result<(), LocaleError> {
-        let key = parse_key(key)?;
+    pub fn set_keyword<S: AsRef<[u8]>>(
+        &mut self,
+        key: S,
+        value: Vec<S>,
```
Comment (emilio) [inline unic-locale-impl/src/extensions/unicode.rs]: Same here.

File: unic-langid-impl/src/lib.rs (hunk 2 above)
Comment (zbraniecki) [reply unic-langid-impl/src/lib.rs]: yeah, I don't think you'd ever want to `from_parts` with just nones.

File: unic-langid-impl/src/lib.rs [hunk 7]
```diff
@@ -142,7 +159,7 @@ impl LanguageIdentifier {
     ///
     /// Not stable.
     pub fn try_from_iter<'a>(
-        iter: &mut Peekable<impl Iterator<Item = &'a str>>,
+        iter: &mut Peekable<impl Iterator<Item = &'a [u8]>>,
```
[the remark this answers is not in the document]
Comment (zbraniecki) [reply unic-langid-impl/src/lib.rs]: yeah, I don't think it is. It's internal really and I'm ok operating over [u8] always.

File: unic-locale-impl/src/extensions/mod.rs (hunk 4 above)
Comment (zbraniecki) [reply unic-locale-impl/src/extensions/mod.rs]: nothing, but the line above guarantees it is at least length of 1 :)

**Pull request merged**